File: src/sftp.py
#!/usr/bin/env python3

"""
Author: Mateus Pinto Garcia
Email: mpgx5.c@gmail.com
Date: 2023-04-18
Co-author: Copilot

A client for connecting to an SFTP server and performing common operations such as
listing directory contents, uploading, and downloading files.

Todas as operações públicas são decoradas com `@safe`, portanto retornam
`Result[..., Exception]` (`Success`/`Failure`) em vez de propagar exceções.
"""

from __future__ import annotations

import logging
from pathlib import Path
from types import TracebackType
from typing import List, Optional

import paramiko
from returns.result import Failure, safe

logger = logging.getLogger(__name__)


class SFTPError(Exception):
    """Erro base do cliente SFTP."""


class NotConnectedError(SFTPError):
    """Levantado quando uma operação é solicitada sem conexão ativa."""


class SFTPClient:
    def __init__(self, hostname: str, port: int, username: str, password: str) -> None:
        """
        Inicializa o cliente SFTP com as credenciais fornecidas.
        """
        self.hostname = hostname
        self.port = port
        self.username = username
        self.password = password
        self.transport: Optional[paramiko.Transport] = None
        self.sftp: Optional[paramiko.SFTPClient] = None

    def __enter__(self) -> "SFTPClient":
        self.connect().unwrap()

        return self

    def __exit__(
        self,
        exc_type: Optional[type[BaseException]],
        exc_val: Optional[BaseException],
        exc_tb: Optional[TracebackType],
    ) -> None:
        result = self.disconnect()

        if isinstance(result, Failure):
            logger.error("Failed to cleanly disconnect from SFTP.")

    def _require_sftp(self) -> paramiko.SFTPClient:
        """
        Devolve a sessão SFTP ativa ou falha se o cliente não estiver conectado.
        """

        if self.sftp is None:
            raise NotConnectedError("SFTP client not connected")

        return self.sftp

    @safe
    def connect(self) -> None:
        """
        Conecta ao servidor SFTP.
        """

        transport = paramiko.Transport((self.hostname, self.port))
        transport.connect(username=self.username, password=self.password)

        sftp = paramiko.SFTPClient.from_transport(transport)

        if sftp is None:
            transport.close()

            raise SFTPError("Could not open an SFTP session over the transport")

        self.transport = transport
        self.sftp = sftp

        logger.info("Connected to SFTP server at %s:%s", self.hostname, self.port)

    @safe
    def disconnect(self) -> None:
        """
        Desconecta do servidor SFTP.
        """

        if self.sftp is not None:
            self.sftp.close()
            self.sftp = None

        if self.transport is not None:
            self.transport.close()
            self.transport = None

        logger.info("Disconnected from SFTP server at %s:%s", self.hostname, self.port)

    @safe
    def listdir(self, remote_path: str) -> List[str]:
        """
        Lista o conteúdo do diretório remoto.
        """

        entries = self._require_sftp().listdir(remote_path)

        logger.info("Listing directory %s: %s", remote_path, entries)

        return entries

    @safe
    def upload(self, local_path: str, remote_path: str) -> None:
        """
        Faz upload de um arquivo local para o caminho remoto.
        """

        source = Path(local_path)

        if not source.exists():
            raise FileNotFoundError(f"Local file {local_path} does not exist.")

        if not source.is_file():
            raise ValueError(f"Local path {local_path} is not a file.")

        self._require_sftp().put(local_path, remote_path)

        logger.info("Uploaded %s to %s", local_path, remote_path)

    @safe
    def download(self, remote_path: str, local_path: str) -> None:
        """
        Faz download de um arquivo remoto para o caminho local.
        """

        self._require_sftp().get(remote_path, local_path)

        logger.info("Downloaded %s to %s", remote_path, local_path)

File: src/ts/sftp.ts
import fs from "node:fs";
import { pipeline } from "node:stream/promises";
import { promisify } from "node:util";
import { Client, SFTPWrapper } from "ssh2";

export type Credentials = {
    host: string;
    port: number;
    username: string;
    password: string;
};

export class SFTPError extends Error {
    constructor(message: string, options?: ErrorOptions) {
        super(message, options);
        this.name = new.target.name;
    }
}

/** Falha no handshake/conexão com o servidor. */
export class SFTPConnectionError extends SFTPError {}

/** Operação solicitada sem uma sessão SFTP ativa. */
export class SFTPNotConnectedError extends SFTPError {}

export class SFTP {
    private client: Client;
    private sftp: SFTPWrapper | null = null;
    private connected: boolean = false;
    private sessionError: Error | null = null;
    private credentials: Credentials;

    constructor({ host, port, username, password }: Credentials) {
        this.client = new Client();

        this.credentials = {
            host,
            port,
            username,
            password,
        };
    }

    async connect(): Promise<void> {
        if (this.connected) {
            throw new SFTPConnectionError("Already connected");
        }

        this.sessionError = null;

        await new Promise<void>((resolve, reject) => {
            const cleanup = () => {
                this.client.off("ready", onReady);
                this.client.off("error", onError);
            };

            const onReady = () => {
                cleanup();
                resolve();
            };

            const onError = (error: Error) => {
                cleanup();
                reject(
                    new SFTPConnectionError(
                        `Failed to connect to ${this.credentials.host}:${this.credentials.port}`,
                        { cause: error },
                    ),
                );
            };

            this.client.once("ready", onReady).once("error", onError);
            this.client.connect(this.credentials);
        });

        // Erros após o handshake viram exceção não tratada do processo se ninguém
        // escutar 'error'; guardamos o último para reportar na próxima operação.
        this.client.on("error", (error: Error) => {
            this.sessionError = error;
            this.connected = false;
            this.sftp = null;
        });

        this.sftp = await promisify(this.client.sftp.bind(this.client))();
        this.connected = true;
    }

    async disconnect(): Promise<void> {
        if (!this.connected) {
            return;
        }

        await new Promise<void>((resolve) => {
            this.client.once("close", () => resolve());
            this.client.end();
        });

        this.connected = false;
        this.sftp = null;
    }

    async upload(source: string, remote: string): Promise<void> {
        const sftp = this.requireSftp();

        await pipeline(fs.createReadStream(source), sftp.createWriteStream(remote));
    }

    async download(remote: string, destination: string): Promise<void> {
        const sftp = this.requireSftp();

        await pipeline(sftp.createReadStream(remote), fs.createWriteStream(destination));
    }

    async listFiles(remoteDir: string): Promise<string[]> {
        const sftp = this.requireSftp();

        const entries = await promisify(sftp.readdir.bind(sftp))(remoteDir);

        return entries.map((entry) => entry.filename);
    }

    private requireSftp(): SFTPWrapper {
        if (this.sessionError) {
            throw new SFTPNotConnectedError("SFTP session was closed by an error", {
                cause: this.sessionError,
            });
        }

        if (!this.sftp) {
            throw new SFTPNotConnectedError("SFTP client not connected");
        }

        return this.sftp;
    }
}
